Document follow_me example

hamishwillee opened the pull request:
This adds the follow me example. It still requires a screenshot.

hamishwillee commented:
@shakthi-prashanth-m I tried to run follow_me example on Ubuntu but got some warnings - what dependencies need to be added?
```
sudo apt-get install libboost*
```

shakthi-prashanth-m commented:
In latest Follow Me `CMakeLists.txt` linking telemetry plugin is missed. So fix is part of QGC Mission import PR. You can temporarily fix it as in [here](https://github.com/dronecore/DroneCore/pull/235/files#diff-ba66a1bdd79144d2f9c22bf9152486b3).
Regarding Dependencies, [Boost version 1.66](https://github.com/dronecore/DroneCore/blob/develop/example/follow_me/CMakeLists.txt#L5) is required.

hamishwillee commented:
@shakthi-prashanth-m Thanks

> Regarding Dependencies, Boost version 1.66 is required.

That is an inconvenient version as Ubuntu 16.04 comes with 1.58. Can the requirement be reduced?

I might just wait until that QGC import goes in (https://github.com/dronecore/DroneCore/pull/235) because realistically people can't use this example now without a lot of hassle.

shakthi-prashanth-m commented:
>That is an inconvenient version as Ubuntu 16.04 comes with 1.58. Can the requirement be reduced?


I had this thought initially, but I chose it to prefer latest Boost version. However, like you said, we can lower the version to match that on Ubuntu 16 for ease of usage. In that case, we have to replace `boost::asio::io_context` by `boost::asio::io_service`.

hamishwillee commented:
>> That is an inconvenient version as Ubuntu 16.04 comes with 1.58. Can the requirement be reduced?

> I had this thought initially, but I chose it to prefer latest Boost version. However, like you said, we can lower the version to match that on Ubuntu 16 for ease of usage. In that case, we have to replace boost::asio::io_context by boost::asio::io_service.

Please do that. Given this is example code it makes sense to make the installation as straightforward as possible (if this were library code then we'd have an argument for most efficient and recent version).

shakthi-prashanth-m commented:
OK. Sure.

shakthi-prashanth-m commented:
@julianoes I think we include fix for Follow Me linker error in this PR which I submit to with Boost version **1.58**, so that @hamishwillee doesn't need to wait for ([dronecore/DroneCore#235](https://github.com/dronecore/DroneCore/pull/235)). I will revert those changes in the other PR.

hamishwillee commented on en/examples/follow_me.md:
```diff
@@ -0,0 +1,341 @@
+This example demonstrates how to use the [Follow Me](../api_reference/classdronecore_1_1_follow_me.md) plugin. 
+It shows how to send the drone both the current position of the target (`FollowMe::TargetLocation`) and the relative position at which it should follow (`FollowMe::Config`).
+
+![Follow Me QGC Screenshot](../../assets/examples/follow_me/follow_me_example_qgc.jpg)
```
@shakthi-prashanth-m I need a screenshot here.

shakthi-prashanth-m replied:
![followme](https://user-images.githubusercontent.com/26615772/36210759-492c1c3a-11c5-11e8-8011-96ae8e1406b9.png)

shakthi-prashanth-m replied:
@hamishwillee I built QGC source code by disabling it sending Follow positions. This made PX4 to receive Follow position only from DroneCore Follow Me example. But Follow Me trajectory is not in a regular shape (I expected this because as I mentioned it as a rudimentary location provider :-) ). So is this fine ?

hamishwillee commented on en/examples/follow_me.md:
````diff
@@ -0,0 +1,341 @@
+> **Note** This is from a debug build of DroneCore. A release build will omit the "Debug" messages.
+
+```
+$ ./follow_me 
````
@shakthi-prashanth-m I need a dronecore console output here.

shakthi-prashanth-m replied:
```
Wait for device to connect via heartbeat
[11:40:49|Info ] New device on: 127.0.0.1:14557 (udp_connection.cpp:211)
[11:40:49|Debug] MAVLink: info: DISARMED by auto disarm on land (device.cpp:247)
[11:40:50|Debug] Discovered 4294967298 (dronecore_impl.cpp:219)
[11:40:50|Info ] FollowMe: Applying default FollowMe configuration FollowMe to the device... (follow_me_impl.cpp:186)
Device is ready
Armed
[11:40:51|Debug] MAVLink: info: ARMED by arm/disarm component command (device.cpp:247)
[11:40:51|Debug] MAVLink: info: [logger] file: rootfs/fs/microsd/log/2018-02-14/0 (device.cpp:247)
In Air...
[11:40:51|Debug] MAVLink: info: Using minimum takeoff altitude: 2.50 m (device.cpp:247)
[11:40:51|Debug] MAVLink: info: Takeoff detected (device.cpp:247)
[11:40:51|Debug] MAVLink: critical: Using minimum takeoff altitude: 2.50 m (device.cpp:247)
[11:40:51|Debug] MAVLink: info: data link #1 lost (device.cpp:247)
[FlightMode: Takeoff] Vehicle is at: nan, nan degrees.
[FlightMode: Hold] Vehicle is at: nan, nan degrees.
[FlightMode: Hold] Vehicle is at: nan, nan degrees.
[FlightMode: Hold] Vehicle is at: nan, nan degrees.
[FlightMode: Hold] Vehicle is at: nan, nan degrees.
[11:40:56|Debug] FollowMe: Waiting for the device confirmation of the new configuration.. (follow_me_impl.cpp:98)
[11:40:56|Debug] FollowMe: Waiting for the device confirmation of the new configuration.. (follow_me_impl.cpp:98)
[11:40:56|Info ] FollowMe: Configured: Min height: 20 meters, Follow distance: 8 meters, Follow direction: Front right, Responsiveness: 0.5 (follow_me_impl.cpp:101)
[FlightMode: FollowMe] Vehicle is at: nan, nan degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54559 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54559 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54563 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54563 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54559 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54559 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54562 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54562 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54566 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54566 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54562 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54562 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54565 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54568 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54568 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54565 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54565 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54567 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54571 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54571 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54567 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54567 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.5457 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.5457 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54567 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54567 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54569 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54566 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54566 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54568 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54565 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54565 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3977, 8.54567 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3976, 8.54567 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3976, 8.5457 degrees.
[FlightMode: FollowMe] Vehicle is at: 47.3976, 8.54573 degrees.
waiting until landed
[11:41:33|Debug] MAVLink: info: Landing at current position (device.cpp:247)
waiting until landed
waiting until landed
waiting until landed
waiting until landed
waiting until landed
waiting until landed
waiting until landed
[11:41:53|Debug] MAVLink: info: data link #1 regained (device.cpp:247)
waiting until landed
waiting until landed
waiting until landed
waiting until landed
waiting until landed
[11:42:04|Debug] MAVLink: info: Landing detected (device.cpp:247)
Landed...
```

hamishwillee commented:
@shakthi-prashanth-m Thanks for your fix https://github.com/dronecore/DroneCore/pull/256. This allowed me to verify the build instructions and run this code.

Unfortunately I'm having trouble getting SITL to run reliably in a VM (lots of failsafe warnings) so I can't get a decent full run of the example. I was wondering if you could do that for me. Basically I need the end to end console output and a screenshot of QGroundControl showing the track that the vehicle took (check out the other examples to see what I mean). Would it be possible for you to generate those for me? The track is also a good example/proof that your location generator is doing what you want.

You can attach the files or email them to me.

shakthi-prashanth-m commented:
@hamishwillee 
If QGroundControl is running when I run Follow Me example, then even it starts sending (Lat, Lan) = (0, 0) too to the vechicle! This makes vehicle move in indeterministic way. So, I thought of running Follow Me example first and then launch QGC to view traveled path; but QGC doesn't show it.

Right now, I can think of disabling sending those position in QGroundControl source code and run by building it. If you've better suggestion, please let me know.

hamishwillee commented:
> @hamishwillee If QGroundControl is running when I run Follow Me example, then even it starts sending (Lat, Lan) = (0, 0) too to the vechicle! This makes vehicle move in indeterministic way. So, I thought of running Follow Me example first and then launch QGC to view traveled path; but QGC doesn't show it.

> Right now, I can think of disabling sending those position in QGroundControl source code and run by building it. If you've better suggestion, please let me know.

OK, so I haven't seen odd behaviour in the other examples - is this a recent build of QGC? That sounds like a bug - QGC shouldn't be sending lat/lon position unless you ask it to. @julianoes Any ideas?

shakthi-prashanth-m commented:
No, It happens in **FollowMe** Mode! Because we enable vehicle to Follo Me Mode, QGroundControl being a client detects it and starts sending Follow positions to vehicle just like DroneCore Follow Me example (in competition :) ) does. So, I think this behavior is normal considering QGC as a client. So, for the time being, just to run Follow Me example we need to disable it in QGC.

hamishwillee commented on en/examples/follow_me.md:
```diff
@@ -0,0 +1,341 @@
+    dronecore
+    dronecore_action
+    dronecore_follow_me
+    pthread
```
Do we really need `pthread` here? Doesn't dronecore link against that?

shakthi-prashanth-m replied:
Yes, its not necessary anymore. Fixed here https://github.com/dronecore/DroneCore/pull/235/commits/91e0a2d325225f8497fd3c4116420a4867f8de18. Thanks @hamishwillee

hamishwillee replied:
Thanks! Corresponding change in docs here: https://github.com/dronecore/docs/commit/a658b68d7362f51e39cf61acb5806fe6af1c408f

hamishwillee commented:
@shakthi-prashanth-m Thanks for your help on this!
1. I have created https://github.com/mavlink/qgroundcontrol/issues/6141 to track the QGC conflict. I've also added a warning in the docs linking to this on Follow Me guide page and in the example. 
2. I added your image and log - thanks for those.

I've merged this now as it is complete.

shakthi-prashanth-m commented:
@hamishwillee I think this code needs to be update to use `follow_distance_m` instead of `follow_dist_m` when you [demonstrate set configuration](https://docs.dronecore.io/v/develop/en/guide/follow_me.html#set-the-follow-configuration).

hamishwillee commented:
Thanks @shakthi-prashanth-m - done in https://github.com/dronecore/docs/commit/26c05b274636d845f4b573b3adedef1c1506029f

Hard to keep track of every single API modifications like that - appreciate your help.
